### PR TITLE
feat: reinvigorate visibleOnFullscreen option

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Refs https://github.com/electron/electron/pull/21706.

Closes https://github.com/electron/electron/issues/24939.

We originally removed this as we could not determine a way to appropriately compensate for Apple's changes to the underlying functionality of NSWindows starting with 10.14 to disallow NSWindows from floating on top of fullscreen apps. At the time we weren't aware of a possible workaround with `app.dock.hide` but now that this is known, and presents value to a nontrivial number of apps, I feel that it would be appropriate to allow apps access to this functionality once more.

Tested on macOS Catalina and observed working as intended.

cc @MarshallOfSound @zcbenz @nornagon 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).
- [x] This is **NOT A BREAKING CHANGE**. Breaking changes may not be merged to master until 11-x-y is branched.

#### Release Notes

Notes: Added back a previously broken `visibleOnFullScreen` option for `setVisibleOnAllWorkspaces`.

### CODE REVIEW

File: shell/browser/native_window_mac.mm [hunk 1]
```diff
@@ -1351,8 +1351,21 @@ void ViewDidMoveToSuperview(NSView* self, SEL _cmd) {
...
                                      const std::string& description) {}
 
-void NativeWindowMac::SetVisibleOnAllWorkspaces(bool visible) {
+void NativeWindowMac::SetVisibleOnAllWorkspaces(bool visible,
+                                                bool visibleOnFullScreen) {
+  // In order for NSWindows to be visible on fullscreen we need to functionally
+  // mimic app.dock.hide() since Apple changed the underlying functionality of
+  // NSWindows starting with 10.14 to disallow NSWindows from floating on top of
+  // fullscreen apps.
+  if (visibleOnFullScreen) {
```
Comment (MarshallOfSound) [inline shell/browser/native_window_mac.mm:1360]: Don't we need to do the inverse of this when `visibleOnFullScreen=false`? i.e. we need an `else` block

**Comment by codebytere** [comment]
Updated to ensure dock behavior restored - tested with https://gist.github.com/1188123a35b35e8c9cd46f34a48ae633

**Comment by codebytere** [comment]
This change is macOS-only, so Window and WOA failures are unrelated.

**Comment by andrewschreiber** [comment]
Any chance we could backport this to electron 8? #24011 is a blocker on upgrading past 8 for now for many of us.

**Comment by ajay-pipaliya-simform** [comment]
> appreciate for bringing this back, however it doesnt seem to work on electron 14 for me
> 
> ```
> win.setAlwaysOnTop(true, 'pop-up-menu');
> win.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true });
> ```
> 
> did it work for you guys?

Nope, if you're using chrome browser in full-screen mode and try to toggle electron app it will show the browserwindow but when you move your cursor towards the browser window it will loss the focus and hide the browser window

**Pull request merged**